### ISSUE TITLE
[OrientDB 2.0.x] StackOverflowError / Infinite Recursion

### ISSUE THREAD

**Issue by Alan47** [orientechnologies/orientdb#4900]
Hello,

I am currently experiencing a StackOverflowError in OrientDB due to an infinite recursion. It happens when I perform a commit on a graph instance. I've tested this issue on 2.0.12 and 2.0.15, with identical results (except for the line numbers).

Unfortunately I cannot provide a reproduction example as the error occurs deep inside the business logic of my application, but here is the repetitive part of the stacktrace (from 2.0.15):

	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeSaveRecord(ODatabaseDocumentTx.java:1693)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.addRecord(OTransactionOptimistic.java:312)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.saveRecord(OTransactionOptimistic.java:253)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:2316)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:118)
	at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1727)
	at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1718)
	at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:76)
	at com.orientechnologies.orient.core.db.record.ridbag.embedded.OEmbeddedRidBag.serialize(OEmbeddedRidBag.java:371)
	at com.orientechnologies.orient.core.db.record.ridbag.ORidBag.toStream(ORidBag.java:274)
	at com.orientechnologies.orient.core.serialization.serializer.record.binary.ORecordSerializerBinaryV0.writeSingleValue(ORecordSerializerBinaryV0.java:583)
	at com.orientechnologies.orient.core.serialization.serializer.record.binary.ORecordSerializerBinaryV0.serialize(ORecordSerializerBinaryV0.java:262)
	at com.orientechnologies.orient.core.serialization.serializer.record.binary.ORecordSerializerBinary.toStream(ORecordSerializerBinary.java:105)
	at com.orientechnologies.orient.core.record.impl.ODocument.toStream(ODocument.java:2100)
	at com.orientechnologies.orient.core.record.impl.ODocument.toStream(ODocument.java:695)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeSaveRecord(ODatabaseDocumentTx.java:1693)

I can consistently reproduce this issue in my codebase. I would be happy if someone could look into this, perhaps it is possible to reconstruct the issue from the stack trace by analyzing the exact call sequence.

**Comment by lvca** [comment]
Could it be that you're saving a big graph? In OrientDB v2.1 the save doesn't use the recursion anymore. Have you already tried with 2.1?

**Comment by Alan47** [comment]
Hi,

the graph I am attempting to store has about 10.000 elements (vertices and edges), so it's not really huge.

I wanted to upgrade to 2.1.x in order to fix the problem, unfortunately I got blocked by this bug:
https://github.com/orientechnologies/spring-data-orientdb/issues/26

**Comment by Alan47** [comment]
One more remark: OrientDB does not get in touch with any Spring or Spring Data class in my project. I just posted my comment in that bug report because the stack trace was almost identical and it seemed to be the same root cause. Please see my comment in the other bug report.

**Comment by Alan47** [comment]
Hello again,

after fixing the issue from the other report, I followed your advice and tried using OrientDB 2.1.1 instead. Unfortunately the situation is still the same. Here's the repetitive part of the stack trace in 2.1.1:

	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeSaveRecord(ODatabaseDocumentTx.java:1830)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.addRecord(OTransactionOptimistic.java:432)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.saveRecord(OTransactionOptimistic.java:363)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:2464)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:120)
	at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1768)
	at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1759)
	at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:81)
	at com.orientechnologies.orient.core.db.record.ridbag.embedded.OEmbeddedRidBag.serialize(OEmbeddedRidBag.java:372)
	at com.orientechnologies.orient.core.db.record.ridbag.ORidBag.toStream(ORidBag.java:266)
	at com.orientechnologies.orient.core.serialization.serializer.record.binary.ORecordSerializerBinaryV0.writeSingleValue(ORecordSerializerBinaryV0.java:579)
	at com.orientechnologies.orient.core.serialization.serializer.record.binary.ORecordSerializerBinaryV0.serialize(ORecordSerializerBinaryV0.java:259)
	at com.orientechnologies.orient.core.serialization.serializer.record.binary.ORecordSerializerBinary.toStream(ORecordSerializerBinary.java:105)
	at com.orientechnologies.orient.core.record.impl.ODocument.toStream(ODocument.java:2285)
	at com.orientechnologies.orient.core.record.impl.ODocument.toStream(ODocument.java:671)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeSaveRecord(ODatabaseDocumentTx.java:1830)

... this looks rather similar to the stacktrace from the earlier version. The exception is thrown at ODatabaseDocumentTx.java:1940:


    com.orientechnologies.orient.core.exception.ODatabaseException: Error on saving record #47:-2986
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeSaveRecord(ODatabaseDocumentTx.java:1940)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.addRecord(OTransactionOptimistic.java:432)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.saveRecord(OTransactionOptimistic.java:363)

I would like to provide more details, but unfortunately all of this code resides very deep down in my application. The graph itself is procedurally generated and different on each test run, so I can't even share the graph structure I'm afraid...

**Comment by shajigopal** [comment]
@Alan47 Any luck on this issue? We are facing the same issue and reported it as #4941 .

**Comment by Alan47** [comment]
No, unfortunately not - I am still facing this problem, and I have no clue how to solve it :(

**Comment by lvca** [comment]
Hi Alan,
Before OrientDB 2.2 (soon in beta), saving used recursive algorithm, so if
you are saving many objects you could incur in this problem. With OrientDB
v2.2 we avoid recursion, so this problem doesn't exist anymore. What to do
with OrientDB v2.0 and v2.1? You have 2 options:

   1. keep the transaction smaller by committing more often or
   2. increase the Java stack size with the -Xss<RAM> option. You could try
   with -Xss1m


Best Regards,

Luca Garulli
Founder & CEO
OrientDB <http://orientdb.com/>

**Comment by Alan47** [comment]
Hi Luca,

thanks for the response. Looking forward to v2.2. Is there any ETA when it will be out? Weeks, months? We will wait for 2.2 and focus on other tasks for now, it would be nice to have some time window for planning purposes.

Kind regards,

Alan

**Comment by tglman** [comment]
Hi,

2.2 has been released, you should not have any problem with big graphs now, closing this.
